src/number_theory/pell.lean

```diff
@@ -137,4 +137,119 @@ end
 
 end existence
 
+/-!
+### Group structure of the solution set
+
+We define a structure of a commutative multiplicative group with distributive negation
+on the set of all solutions to the Pell equation `x^2 - d*^2 = 1`.
+
+The type of such solutions is `pell.solution₁ d`. It contains integers `x` and `y` and
+a proof that `(x, y)` is indeed a solution.
+
+The multiplication is given by `(x, y) * (x', y') = (x*y' + d*y*y', x*y' + y*x')`
+(this is obtained by mapping `(x, y)` to `x + y*√d` and multiplying the results).
```

Comment:
Is there an isomorphism between `solution d 1` and `units (zsqrt d)`?

Reply:
There is an embedding into `units (zsqrtd d)`, which is in general not surjective. `unitary ...` would seem to be the correct thing, but `zsqrtd d` has no `star_semigroup` instance (and neither has `units (zsqrtd d)`.
In any case, I am planning to define the map and prove its relevant properties at some point, but it is not (yet) needed for the result on the structure of the solutions.

Reply:
I wonder if it's more valuable to just
* Provide the `star_ring` instance
* Add `x ∈ unitary (zsqrt d) ↔ x.re^2 - d*x.im^2 = 1`
* Optionally define `solutions₁= unitary (zsqrt d)`

It's not clear to me what the benefit of building the solution space from scratch is when we can add some missing glue to build it trivially

Reply:
One consideration is that this will not work so well with `solution d m` (which has an action by `solution d 1`).
But I'll play around with this a bit to see whether it would be a viable alternative.

Reply:
If we go in this direction, I would suggest to add the following to `number_theory.zsqrtd.basic`:
```lean
namespace zsqrtd

variables {d : ℤ}

instance : has_star (zsqrtd d) := { star := conj }

instance : star_semigroup  (zsqrtd d) :=
{ star := has_star.star,
  star_involutive := λ a, by simp only [star, conj_conj],
  star_mul := λ a b, by simp only [star, conj_mul, mul_comm] }

instance : star_ring (zsqrtd d) :=
{ to_star_semigroup := infer_instance,
  star_add := λ a b, by simp only [star, conj_add] }

@[ext] lemma ext' {a b : zsqrtd d} (hre : a.re = b.re) (him : a.im = b.im) : a = b :=
ext.mpr ⟨hre, him⟩

lemma star_mul_self_eq_norm (a : zsqrtd d) : star a * a = norm a :=
by { ext; simp [star, norm, conj, coe_int_val]; ring }

lemma self_mul_star_eq_norm (a : zsqrtd d) : a * star a = norm a :=
by { rw mul_comm, exact star_mul_self_eq_norm a }

lemma norm_eq_one_iff_mem_unitary {a : zsqrtd d} :
  a.re ^ 2 - d * a.im ^ 2 = 1 ↔ a ∈ unitary (zsqrtd d) :=
begin
  rw [unitary.mem_iff, star_mul_self_eq_norm, self_mul_star_eq_norm, and_self, norm_def,
      sq, sq, ← mul_assoc],
  norm_cast,
end

end zsqrtd
```lean
(A suitable `ext` lemma was also missing...)

Then we can do the following.
```lean
@[derive [group, has_distrib_neg, inhabited]]
def solution₁ (d : ℤ) : Type := ↥(unitary (zsqrtd d))

namespace solution₁

variables {d : ℤ}

instance : has_coe (solution₁ d) (zsqrtd d) := { coe := subtype.val } 

def x (a : solution₁ d) : ℤ := (a : zsqrtd d).re

def y (a : solution₁ d) : ℤ := (a : zsqrtd d).im

lemma rel (a : solution₁ d) : a.x ^ 2 - d * a.y ^ 2 = 1 :=
by exact_mod_cast zsqrtd.norm_eq_one_iff_mem_unitary.mpr a.prop

@[ext]
lemma ext {a b : solution₁ d} (hx : a.x = b.x) (hy : a.y = b.y) : a = b :=
by {ext; assumption}

def mk (x y : ℤ) (rel : x ^ 2 - d * y ^ 2 = 1) : solution₁ d :=
{ val := ⟨x, y⟩,
  property := zsqrtd.norm_eq_one_iff_mem_unitary.mp rel }

@[simp]
lemma x_mk (x y : ℤ) (rel : x ^ 2 - d * y ^ 2 = 1) : (mk x y rel).x = x := rfl

@[simp]
lemma y_mk (x y : ℤ) (rel : x ^ 2 - d * y ^ 2 = 1) : (mk x y rel).y = y := rfl
```
Then one can define `rel_x`, `rel_y` and all the `simp` lemmas essentially as before, but does not have to bother with defining the various instances. Then everything that comes later (but is not yet in this PR) works with almost no changes.

So it looks like it makes sense to do it in this way. I'll wait for more comments, and if no modifications are suggested, I'll make the change.

Reply:
I've made #18572 to switch from `zsqrtd.conj` to `star`

Reply:
OK, thanks.
I'll then wait until this is merged, so that I can build on it.